ENH: Add parsing for ISO_FORTRAN_ENV to F2Py.

DWesl opened the pull request:
ISO_FORTRAN_ENV has KINDs for INT8-INT64 and REAL32-REAL128. Not all are guaranteed to be present on all platforms, even with GCC. I think equivalents for the ones I included here are required by the C standards, and most Fortran compilers are paired with a C compiler, so these should be present.

The addition mostly follows the pattern of the ISO_C_BINDING module already supported by F2Py.

<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

DWesl committed:
ENH: Add parsing for ISO_FORTRAN_ENV to F2Py.

ISO_FORTRAN_ENV has KINDs for INT8-INT64 and REAL32-REAL128.
Not all are guaranteed to be present on all platforms, even with GCC.
I think equivalents for the ones I included here are required by the C standards,
and most Fortran compilers are paired with a C compiler, so these should be present.

The addition mostly follows the pattern of the
ISO_C_BINDING module already supported by F2Py.

FIX:CI:F2Py Fix typos in test code

STY,FIX: Ensure arguments to float functions are floats

Maybe this'll help with the real64 addition?

jorenham commented on numpy/f2py/_isofenv.py:
```diff
@@ -0,0 +1,23 @@
+iso_c2py_map = {}
+
+isof_kindmap = {}
+for fortran_type, std_type_dict in iso_fortran_env_map.items():
+    for std_type in std_type_dict.keys():
+        isof_kindmap[std_type] = fortran_type
```
These local variables will cause some namespace pollution. And because there's no `__all__`, stubtest is probably going to require stub annotations for e.g. `std_type`.  

You can see some examples of that in #28522 in the places I marked with `# namespace pollution`. So it'd help we can prevent it this time :)

DWesl committed:
TYP: Add type annotations to used variables

Also mark private variables and delete unused variables.

DWesl committed:
TST:F2Py: Add equivalent tests to ISO C test module.

Let's see if the ISO C bindings can handle 2**32 passed to a C int64_t

DWesl committed:
ENH:F2Py: Add ISO_FORTRAN_ENVIRONMENT specifiers to capi maps

Let's see if this lets the tests pass

DWesl committed:
TST:F2Py: Add int8 test to ISO C binding

Try to work out why the new ISO_FORTRAN_ENV tests fail

DWesl commented:
I think I have int64 working, but int8 is still broken, and doesn't seem to be working with ISO_C_BINDING either

DWesl committed:
BUG: Move iso_c_binding use inside submodule using it

Ran into problems testing on gfortran: this shift ameliorated those issues and allowed parsing as valid Fortran

BUG: Move module for type kinds into each subroutine

This seems to be what Fortran scoping rules require

mattip commented:
Needs a rebase

DWesl committed:
STY: Include rust suggestions

DWesl committed:
TST: Do not specify intrinsic module.

F2Py does not seem to recognize that construct.

charris commented:
@DWesl do you want to keep working on this?

DWesl committed:
TST: Drop int8 ISO-F-env tests.

Apparently they work for ISO C binding, but I just get errors for ISO F env.

DWesl commented:
I got the tests passing locally again by giving up on int8; the approach used for ISO C binding just gives me errors about 'signed_char not a C type'

This required reverting the Cygwin skip from #25111, which doesn't seem to be needed anymore.

(I get segfaults in `numpy._core.tests.test_dtype::TestMonsterType`, so Cygwin support still needs work in general)

DWesl committed:
TST: Drop ISO C BINDING int8 test.

This doesn't work either.

DWesl committed:
TST: Revert extra changes to F2Py ISO C Binding tests.

These weren't related to the core PR, and were added for debugging the int8/signed_char difficulties.

DWesl committed:
TYP: Add type stub for new function.

DWesl commented:
Right, that splits the int8 work into a later project, giving up on getting that working with the ISO C Binding or ISO Fortran Environment modules here.  Do you want me to smoosh that down to one commit as a rebase?

charris commented:
> Do you want me to smoosh that down to one commit as a rebase?

If you think that clarifies things, otherwise I will just squash and merge.

DWesl commented:
> > Do you want me to smoosh that down to one commit as a rebase?
> 
> If you think that clarifies things, otherwise I will just squash and merge.

That works too

charris commented:
Thanks @DWesl .
